app/scripts/lib/transaction/util.ts

```diff
@@ -0,0 +1,190 @@
+  userOperationController: UserOperationController;
+};
+
+type FinalAddTransactionRequest = BaseAddTransactionRequest & {
+  transactionOptions: AddTransactionOptions;
+};
+
+export type AddTransactionRequest = FinalAddTransactionRequest & {
+  waitForSubmit: boolean;
+};
+
+export type AddDappTransactionRequest = BaseAddTransactionRequest & {
+  dappRequest: Record<string, any>;
+};
+
+export async function addDappTransaction(
+  request: AddDappTransactionRequest,
+): Promise<string> {
+  const { dappRequest } = request;
+  const { id: actionId, method, origin } = dappRequest;
+
+  ///: BEGIN:ONLY_INCLUDE_IF(blockaid)
+  const { securityAlertResponse } = dappRequest;
+  ///: END:ONLY_INCLUDE_IF
+
+  const transactionOptions: AddTransactionOptions = {
+    actionId,
+    method,
+    origin,
+    // This is the default behaviour but specified here for clarity
+    requireApproval: true,
+    ///: BEGIN:ONLY_INCLUDE_IF(blockaid)
+    securityAlertResponse,
+    ///: END:ONLY_INCLUDE_IF
+  };
+
+  const { waitForHash } = await addTransactionOrUserOperation({
+    ...request,
+    transactionOptions,
+  });
+
+  return (await waitForHash()) as string;
+}
+
+export async function addTransaction(
+  request: AddTransactionRequest,
+): Promise<TransactionMeta> {
+  const { waitForSubmit } = request;
+
+  const { transactionMeta, waitForHash } = await addTransactionOrUserOperation(
+    request,
+  );
+
+  if (!waitForSubmit) {
+    waitForHash().catch(() => {
+      // Not concerned with result.
+    });
+
+    return transactionMeta as TransactionMeta;
+  }
+
+  const transactionHash = await waitForHash();
+
+  const finalTransactionMeta = getTransactionByHash(
+    transactionHash as string,
+    request.transactionController,
+  );
+
+  return finalTransactionMeta as TransactionMeta;
+}
+
+async function addTransactionOrUserOperation(
+  request: FinalAddTransactionRequest,
+) {
+  const { selectedAccount } = request;
+
+  const isSmartContractAccount = selectedAccount.type === 'eip155:eip4337';
+
+  if (isSmartContractAccount) {
+    return addUserOperationWithController(request);
+  }
+
+  return addTransactionWithController(request);
+}
+
+async function addTransactionWithController(
+  request: FinalAddTransactionRequest,
+) {
+  const { transactionController, transactionOptions, transactionParams } =
+    request;
+
+  const { result, transactionMeta } =
+    await transactionController.addTransaction(
+      transactionParams,
+      transactionOptions,
+    );
+
+  return {
+    transactionMeta,
+    waitForHash: () => result,
+  };
+}
+
+async function addUserOperationWithController(
+  request: FinalAddTransactionRequest,
+) {
+  const {
+    networkClientId,
+    transactionController,
+    transactionOptions,
+    transactionParams,
+    userOperationController,
+  } = request;
+
+  const { maxFeePerGas, maxPriorityFeePerGas } = transactionParams;
+  const { origin, requireApproval, type } = transactionOptions as any;
```

Comment:
We might not need to cast it here. Unless we need something down the line.

Reply:
This is intentional for now as the `SmartContractAccount` at this point is required but I have a PR pending to make it optional and use the snap account by default.